File: go/24_tree/StackBasedOnArray.go
package _4_tree

import "fmt"

/*
基于数组实现的栈
*/

type ArrayStack struct {
	//数据
	data []interface{}
	//栈顶指针
	top int
}

func NewArrayStack() *ArrayStack {
	return &ArrayStack{
		data: make([]interface{}, 0, 32),
		top:  -1,
	}
}

func (this *ArrayStack) IsEmpty() bool {
	if this.top < 0 {
		return true
	}
	return false
}

func (this *ArrayStack) Push(v interface{}) {
	if this.top < 0 {
		this.top = 0
	} else {
		this.top += 1
	}

	if this.top > len(this.data)-1 {
		this.data = append(this.data, v)
	} else {
		this.data[this.top] = v
	}
}

func (this *ArrayStack) Pop() interface{} {
	if this.IsEmpty() {
		return nil
	}
	v := this.data[this.top]
	this.top -= 1
	return v
}

func (this *ArrayStack) Top() interface{} {
	if this.IsEmpty() {
		return nil
	}
	return this.data[this.top]
}

func (this *ArrayStack) Flush() {
	this.top = -1
}

func (this *ArrayStack) Print() {
	if this.IsEmpty() {
		fmt.Println("empty statck")
	} else {
		for i := this.top; i >= 0; i-- {
			fmt.Println(this.data[i])
		}
	}
}

File: rust/08_stack/stack_based_on_array.rs
#[derive(Debug)]
struct ArrayStack {
    data: Vec<i32>,
    top: i32,
}

impl ArrayStack {
    fn new() -> Self {
        ArrayStack { data: Vec::with_capacity(32), top: -1 }
    }

    fn push(&mut self, x: i32) {
        self.top += 1;
        if self.top == self.data.capacity() as i32 {
            let tmp_arr = self.data.clone();
            self.data = Vec::with_capacity(self.data.capacity() * 2);
            for d in tmp_arr.into_iter() {
                self.data.push(d);
            }
        }
        self.data.push(x);
    }

    fn pop(&mut self) {
        if self.is_empty() { return; }
        self.top -= 1;
        self.data.pop();
    }

    fn top(&self) -> i32 {
        if self.is_empty() { return -1; }
        *self.data.last().unwrap()
    }

    fn is_empty(&self) -> bool {
        if self.top < 0 { true } else { false }
    }
}

fn main() {
    let mut stack = ArrayStack::new();
    stack.push(-2);
    stack.push(0);
    stack.push(-3);
    stack.pop();
    println!("{:?}", stack.top());
}
